_posts/2019-03-20-dask-cuml.md

```diff
@@ -0,0 +1,235 @@
+```
+
+When we started experimenting with hyper-parameter optimization, we found a few API holes missing, these were
+resolved, mostly handling matching argument structure and various getters/setters.
+
+- [#271](https://github.com/rapidsai/cuml/pull/271)
+- [#318](https://github.com/rapidsai/cuml/pull/318)
+- [#330](https://github.com/rapidsai/cuml/pull/330)
+- [#322](https://github.com/rapidsai/cuml/pull/322)
+
+With holes plugged up we tested again.  Using the same diabetes data set, we are now performing hyper-parameter optimization
+and searching over many alpha parameters for the best *scoring* alpha.
+
+```python
+params = {'alpha': np.logspace(-3, -1, 10)}
+clf = linear_model.Ridge(alpha=alpha, fit_intercept=fit_intercept, normalize=normalize, solver='cholesky')
+cu_clf = cumlRidge(alpha=alpha, fit_intercept=fit_intercept, normalize=normalize, solver="eig")
+
+grid = GridSearchCV(clf, params, scoring='r2')
+grid.fit(X_train, y_train)
+
+cu_grid = GridSearchCV(cu_clf, params, scoring='r2')
+cu_grid.fit(X_train, y_train)
+```
+
+Again, reminding ourselves that the data is small ~28KB, we don't expect to observe cuml performing faster than sklearn. Instead, we want to demonstrate functionality.  Fortunately, GridSearchCV from dask-ml adheres to the same API as sklearn; we can swap out GridsearchCV from sklearn and use dask-ml's which will use all of the GPUs we have available.
+
+```python
+params = {'alpha': np.logspace(-3, -1, 10)}
+clf = linear_model.Ridge(alpha=alpha, fit_intercept=fit_intercept, normalize=normalize, solver='cholesky')
+cu_clf = cumlRidge(alpha=alpha, fit_intercept=fit_intercept, normalize=normalize, solver="eig")
+
+grid = dcv.GridSearchCV(clf, params, scoring='r2')
+grid.fit(X_train, y_train)
+
+cu_grid = dcv.GridSearchCV(cu_clf, params, scoring='r2')
+cu_grid.fit(X_train, y_train)
+```
+
+Timing Measurements:
+
+| | sklearn | cuml |
+| -----------| --- | -----|
+| **GridSearchCV** | 88.4 ms ± 6.11 ms | 6.51 s ± 132 |
+| **dask-ml GridSearchCV** | 873 ms ± 347| 740 ms ± 142 |
+
+
+Unsurprisingly, we see that GridSearchCV and Ridge Regression from sklearn is the fastest in this context.
+There is cost to distributing work and data, and as we previously mentioned, moving data from host to device.
+
+### How does performance scale as we scale data?
+
+```python
+two_dup_data = np.array(np.vstack([X_train]*int(1e2)))
+two_dup_train = np.array(np.hstack([y_train]*int(1e2)))
+three_dup_data = np.array(np.vstack([X_train]*int(1e3)))
+three_dup_train = np.array(np.hstack([y_train]*int(1e3)))
+
+cu_grid = dcv.GridSearchCV(cu_clf, params, scoring='r2')
+cu_grid.fit(two_dup_data, two_dup_train)
+
+cu_grid = dcv.GridSearchCV(cu_clf, params, scoring='r2')
+cu_grid.fit(three_dup_data, three_dup_train)
+
+grid = dcv.GridSearchCV(clf, params, scoring='r2')
+grid.fit(three_dup_data, three_dup_train)
+```
+
+Timing Measurements:
+
+- cuml + dask-ml (2.8MB): 1min 17s
+- cuml + dask-ml (28MB): 13.8 s
+- sklearn + dask-ml: 4.87 s
```

Comment:
My brain originally expected the same table as above, but now with a larger dataset.  This is good information though too.